#41 Avoid issues when the package is scheduled for removal

Jean85 opened the pull request:
This is to solve #41, as proposed in [my comment](https://github.com/Ocramius/PackageVersions/issues/41#issuecomment-326892408).

Ocramius commented on src/PackageVersions/NotInstalledException.php:
```diff
@@ -0,0 +1,15 @@
+
+class NotInstalledException extends \RuntimeException
+{
+    const MESSAGE = "Folder of ocramius/package-versions not found, it seems that it's not installed";
```
Move to a static named constructor instead

Ocramius commented on src/PackageVersions/Installer.php:
```diff
@@ -92,11 +92,17 @@ public static function dumpVersionsClass(Event $composerEvent)
+                $composer->getConfig(),
+                $composer->getPackage()
+            );
+        } catch (NotInstalledException $exception) {
```
Since you are using private-only API, throwing and catching in the same scope seems like a bad idea (it's basically a GOTO implemented via exceptions).

Can this be re-designed somehow?

Jean85 commented:
I've addressed your concerns in two different commits, so it should be easy to review:

 * I've refactored to avoid the exception altogether
 * I've added a step on the Travis build that does an integration test with composer to avoid regressions on this issue

Jean85 commented:
Ok now build is green! @Ocramius if it's :+1: for you, it's ready to be merged!

For some unknown reason tests were failing due to the presence of the new composer.json for the integration test. I preferred switching to a new E2E test for this scenario, it works well too.

Ocramius commented:
@Jean85 merged, thanks! :+1:

Jean85 commented:
You're welcome! :wink:

ondrejmirtes commented:
@Ocramius @Jean85 Can you both release new versions of your packages so I can require them as minimal in PHPStan's dependencies? :) Thanks.

Ocramius commented:
@ondrejmirtes this was just released: https://github.com/Ocramius/PackageVersions/releases/tag/1.1.3

ondrejmirtes commented:
Good, now @Jean85 should require this version in `jean85/pretty-package-versions` and release a new version :)

ondrejmirtes commented:
Good, now @Jean85 should require this version in `jean85/pretty-package-versions` and release a new version :) And then I will release PHPStan 0.8.5 :)

Ocramius commented:
@ondrejmirtes given that `jean85/pretty-package-versions` works as-is, does @Jean85 really need to make a release? Even with preferred stable dependencies, the last patch version should be allowed.

ondrejmirtes commented:
@Ocramius I could add a constraint for your package so that it’s resolved
to a correct version but I’d prefer if there was a correct version of
jean85/pretty-package-versions
so that t works without me caring about indirect dependencies.

On Wed, 6 Sep 2017 at 17:33, Marco Pivetta <notifications@github.com> wrote:

> @ondrejmirtes <https://github.com/ondrejmirtes> given that
> jean85/pretty-package-versions works as-is, does @Jean85
> <https://github.com/jean85> really need to make a release? Even with
> preferred stable dependencies, the last patch version should be allowed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ocramius/PackageVersions/pull/46#issuecomment-327522575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGZuJtxGb3Sj8poAuLa6srDc_-zxl9Mks5sfrtHgaJpZM4PLvxR>
> .
>
-- 

Ondřej Mirtes

Jean85 commented:
No biggies, I can release a patch version too. Doing it right now...
